astroid/nodes/node_classes.py

```diff
@@ -2020,7 +2023,7 @@ def postinit(self, nodes: typing.List[NodeNG]) -> None:
         """
         self.nodes = nodes
 
-    def scope(self):
+    def scope(self) -> Optional["LocalsDictNodeNG"]:
```

Comment:
I was debating with myself what to do here. Technically, `Optional` is correct as nodes can be created without parents. In practice however, only `Module` nodes won't have a `parent`.

We could remove the `Optional` and add `type: ignore` to the function body. The drawback is that we risk missing some obscure corner case where it's actually `None`. The safest course of action would likely be to leave it and modify `pylint` with `node.scrope() is not None` checks.

Reply:
Just thinking out loud here, but what do we currently do when `scope()` returns `None`? As in, what is the use case of that option? I can imagine in those cases we want to go back a level and get the scope of the module. Or are there are checks that depend on the fact that `scope()` is `None`?
If the latter is false we might do something were if `scope()` would return `None` and `isinstance(self, nodes.Module)` we return self. And if it's not a Module do some error-handling. But this is completely dependent on the question whether there are any actual use-cases for returning `None`

Reply:
> Just thinking out loud here, but what do we currently do when `scope()` returns `None`? As in, what is the use case of that option?

We can't really do anything. It's only important to make sure the it doesn't create an error. E.g. by trying to access an attribute of it.

> But this is completely dependent on the question whether there are any actual use-cases for returning `None`

As `None` shouldn't be returned unless someone didn't properly assign `parent`, the only thing for us to do in these cases is to ignore the node in the current check (i.e. `return` early) and move on.

--
**OR** we could decide that we actually want to fail in these cases (as they are errors anyway). In which case we could remove `Optional` and add `type: ignore` as explained earlier.

Reply:
I would be in favour of the latter. I feel adding checks to the `pylint` codebase for `scope() != None` should not be the preferred solution as it delegates an `astroid` problem to `pylint`. IMO a solution within `astroid` is preferable.

If I understand you correctly you want to leave the function body as is and just add a `type: ignore`. You probably have good reasons for this (as you always do 😄 ), but why wouldn't you want to check for `self.parent` and raise an `AttributeError` when it is not assigned?

Reply:
> If I understand you correctly you want to leave the function body as is and just add a `type: ignore`. You probably have good reasons for this (as you always do 😄 )

Actually I don't have a reason for that specifically 😅

> but why wouldn't you want to check for `self.parent` and raise an `AttributeError` when it is not assigned?

Raising an exception is a good idea! Maybe not `AttributeError`, but a custom one inheriting from `AstroidError`. We could include the name of the class that doesn't have a parent in the error message. That should speed up debugging.